File: genlinks.rb
#! ruby

require 'csv'
require 'erb'
require 'optparse'


SCRIPT_VERSION = "v0.1.0"

class LinkGenerator

  TEMPLATE = <<EOT
<html>
  <body>
    <ol>
      <%- @links.each do |link| -%>
      <li><a href="<%= link.url %>"><%= link.name || link.url %></a></li>
      <%- end -%>
    </ol>
  </body>
</html>
EOT


  def initialize(file)
    @links = File.readlines(file).map do |l|
      url, name, banner = l.parse_csv
      Link.new(url, name, banner)
    end
  end

  def to_html(template_file = nil)
    if template_file
      template = File.read(template_file)
    else
      template = TEMPLATE
    end
    erb = ERB.new(template, nil, "-")
    erb.result(binding)
  end


  Link = Struct.new(:url, :name, :banner)


end   # of class LinkGenerator



options = {}
opt = OptionParser.new
opt.banner = <<EOB
Usage: #{opt.program_name} [options] <file>
EOB
opt.on('-t', '--template=FILE', 'Use FILE as template.'){|v|
  options[:template] = v
}
opt.on_tail('-v', '--version', 'Show version.'){|v|
  print SCRIPT_VERSION + "\n"
  exit
}
opt.on_tail('-h', '--help', 'Show this message.'){|v|
  print opt
  exit
}
opt.parse!


print LinkGenerator.new(ARGV.shift).to_html(options[:template])

File: genlinks.pl
use strict;
use warnings;

use Text::CSV_XS;
use HTML::Template;
use Getopt::Long;


my $script_version = "v0.1.0";

sub usage {
    print "Usage: genlinks.pl [options] <file>\n";
    print "Options:\n";
    print "    --version         Show version.\n";
    print "    --help            Show this massage.\n";
}

my $opt_version = 0;
my $opt_help    = 0;
GetOptions('version' => \$opt_version, 'help' => \$opt_help);

if ($opt_help) {
  usage;
  exit;
} elsif ($opt_version) {
  print "$script_version\n";
  exit;
}


my $src =<<EOT;
<html>
  <body>
    <ol>
      <TMPL_LOOP NAME=LINKS>
      <li><a href="<TMPL_VAR NAME=URL>">
        <TMPL_IF NAME=NAME>
          <TMPL_VAR NAME=NAME>
        <TMPL_ELSE>
          <TMPL_VAR NAME=URL>
        </TMPL_IF>
      </a></li>
      </TMPL_LOOP>
    </ol>
  </body>
</html>
EOT

my $filename = shift @ARGV;
open(my $fh, '<', $filename) or die  "Cannot open file: $filename\n";

my @links;
my $tc = Text::CSV_XS->new;
while(<$fh>){
  next unless $tc->parse($_);
  my @fields = $tc->fields;
#  push(@links, { URL => $fields[0], NAME => $fields[1], BANNER => $fields[2]});
  push(@links, { URL => $fields[0], NAME => $fields[1]});
}

my $template = HTML::Template->new(scalarref => \$src);
$template->param(LINKS => \@links);
print $template->output;


